Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt

```diff
@@ -1,34 +1,156 @@
+                .setPayload(ppaData)
+                .build()
+
+            dataDonationAnalyticsServer.uploadAnalyticsData(ppaContainer)
+
+            return true
+        } catch (err: Exception) {
+            Timber.e(err, "Error during analytics submission")
+            return false
+        }
+    }
+
+    suspend fun collectContributions(ppaDataBuilder: PpaData.PPADataAndroid.Builder): List<DonorModule.Contribution> {
+        val request: DonorModule.Request = object : DonorModule.Request {}
+
+        val contributions = donorModules.map {
+            Timber.d("Beginning donation for module: %s", it::class.simpleName)
+            it.beginDonation(request)
+        }
+
+        contributions.forEach {
+            Timber.d("Injecting contribution: %s", it::class.simpleName)
+            it.injectData(ppaDataBuilder)
+        }
+
+        return contributions
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     suspend fun submitAnalyticsData() {
-//        val request: DonorModule.Request = ...
-//        // Collect data from all donor modules
-//        val contributions = donorModules.map { it.beginDonation(request) }
-//        val ppaContainer: Any = ...
-//
-//        contributions.forEach {
-//            it.injectData(ppaContainer)
-//        }
-//
-//        val success = trySubmission(ppaContainer)
-//        contributions.forEach {
-//            it.finishDonation(success)
-//        }
+        Timber.d("Starting analytics submission")
+
+        val ppaDataBuilder = PpaData.PPADataAndroid.newBuilder()
+
+        val contributions = collectContributions(ppaDataBuilder = ppaDataBuilder)
+
+        val analyticsProto = ppaDataBuilder.build()
+
+        val success = trySubmission(analyticsProto)
+
+        contributions.forEach {
+            Timber.d("Finishing contribution: %s", it::class.simpleName)
+            it.finishDonation(success)
+        }
+
+        if (success) {
+            settings.lastSubmittedTimestamp.update {
+                Instant.now()
+            }
+
+            logger.storeAnalyticsData(analyticsProto)
+        }
+
+        Timber.d("Finished analytics submission success=%s", success)
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToNoUserConsent(): Boolean {
+        return !settings.analyticsEnabled.value
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    suspend fun stopDueToProbabilityToSubmit(): Boolean {
+        val submitRoll = Random.nextDouble(0.0, 1.0)
+        return submitRoll > appConfigProvider.getAppConfig().analytics.probabilityToSubmit
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToLastSubmittedTimestamp(): Boolean {
+        val lastSubmit = settings.lastSubmittedTimestamp.value ?: return false
+        return lastSubmit.plus(Hours.hours(LAST_SUBMISSION_MIN_AGE_HOURS).toStandardDuration()).isAfterNow
```

Comment:
`isAfterNow` internally calls `Instant.now` right? `TimeStamper` so we can fake this in tests?